File: apps/translate-tencent_HY-MT/Android/app/src/main/java/ai/zetic/demo/offlinetranslator/viewmodel/TranslationViewModel.kt
package ai.zetic.demo.offlinetranslator.viewmodel

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import ai.zetic.demo.offlinetranslator.model.Language
import ai.zetic.demo.offlinetranslator.ocr.ImageTextRecognizer
import ai.zetic.demo.offlinetranslator.service.NetworkMonitor
import ai.zetic.demo.offlinetranslator.service.SpeechController
import ai.zetic.demo.offlinetranslator.service.VoiceInputController
import ai.zetic.demo.offlinetranslator.translation.TranslationCleanup
import ai.zetic.demo.offlinetranslator.translation.TranslationPrompt
import ai.zetic.demo.offlinetranslator.translation.Translator
import ai.zetic.demo.offlinetranslator.translation.TranslatorFactory
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Locale
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

sealed interface ModelState {
    data object Loading : ModelState
    data object Ready : ModelState
    data class Failed(val message: String) : ModelState
}

/**
 * Streaming translation state + on-device engine driver. Faithful Kotlin port of iOS
 * `TranslationViewModel`.
 *
 * Threading (critical): the ZeticMLange native model is blocking and NOT thread-safe — load, run,
 * waitForNextToken and cleanUp must all happen on ONE dedicated thread, never a thread pool (a
 * coroutine IO/Default dispatcher would crash native init). A single-thread executor serializes
 * every engine call. [gen] holds the latest generation id so an in-flight translation stops early
 * when superseded (mirrors the iOS `genLock`).
 */
class TranslationViewModel(app: Application) : AndroidViewModel(app) {

    // Editable translation state
    private val _sourceText = MutableStateFlow("")
    val sourceText: StateFlow<String> = _sourceText.asStateFlow()
    private val _translatedText = MutableStateFlow("")
    val translatedText: StateFlow<String> = _translatedText.asStateFlow()
    private val _source = MutableStateFlow(Language.detect)
    val source: StateFlow<Language> = _source.asStateFlow()
    private val _target = MutableStateFlow(Language.named("en"))
    val target: StateFlow<Language> = _target.asStateFlow()

    // Model + generation state
    private val _modelState = MutableStateFlow<ModelState>(ModelState.Loading)
    val modelState: StateFlow<ModelState> = _modelState.asStateFlow()
    private val _downloadProgress = MutableStateFlow(0.0)
    val downloadProgress: StateFlow<Double> = _downloadProgress.asStateFlow()
    private val _isTranslating = MutableStateFlow(false)
    val isTranslating: StateFlow<Boolean> = _isTranslating.asStateFlow()

    // Voice / OCR input state
    private val _isListening = MutableStateFlow(false)
    val isListening: StateFlow<Boolean> = _isListening.asStateFlow()
    private val _partialVoiceText = MutableStateFlow("")
    val partialVoiceText: StateFlow<String> = _partialVoiceText.asStateFlow()
    private val _isRecognizingImage = MutableStateFlow(false)
    val isRecognizingImage: StateFlow<Boolean> = _isRecognizingImage.asStateFlow()
    private val _inputError = MutableStateFlow<String?>(null)
    val inputError: StateFlow<String?> = _inputError.asStateFlow()
    // One-shot signal: text captured from voice/OCR → screen should show the Result phase.
    private val _showResult = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val showResult: SharedFlow<Unit> = _showResult.asSharedFlow()

    private val translator: Translator = TranslatorFactory.create(app)
    private val speech = SpeechController(app)
    private val voice = VoiceInputController(viewModelScope)
    private val networkMonitor = NetworkMonitor(app)
    val isOnline: StateFlow<Boolean> get() = networkMonitor.isOnline

    // ONE dedicated engine thread for every native call. Never a coroutine pool.
    private val engine = Executors.newSingleThreadExecutor { r ->
        Thread(r, "zetic-engine").apply { priority = Thread.NORM_PRIORITY + 1 }
    }
    private val gen = AtomicInteger(0)

    init {
        networkMonitor.start()
        loadModel()
    }

    // MARK: - Model lifecycle

    fun loadModel() {
        _modelState.value = ModelState.Loading
        _downloadProgress.value = 0.0
        Log.i(TAG, "Loading model ${translator::class.simpleName} …")
        engine.execute {
            try {
                translator.load { progress -> _downloadProgress.value = progress }
                Log.i(TAG, "Model ready")
                _modelState.value = ModelState.Ready
            } catch (t: Throwable) {
                Log.e(TAG, "Model load failed", t)
                _modelState.value = ModelState.Failed(t.message ?: "Failed to load the model.")
            }
        }
    }

    // MARK: - Translation

    fun canTranslate(): Boolean =
        _modelState.value == ModelState.Ready && _sourceText.value.trim().isNotEmpty()

    fun translate() {
        val text = _sourceText.value.trim()
        if (text.isEmpty() || _modelState.value != ModelState.Ready) return

        // New generation id supersedes any in-flight one (its callback sees a newer id and stops).
        val myGen = gen.incrementAndGet()
        _translatedText.value = ""
        _isTranslating.value = true

        val prompt = TranslationPrompt.make(text, _source.value, _target.value)
        engine.execute {
            translator.reset() // clear KV/conversation state from any prior turn
            val streamed = StringBuilder()
            try {
                translator.generate(prompt) { token ->
                    if (gen.get() != myGen) return@generate false // superseded → stop
                    streamed.append(token)
                    val snapshot = TranslationCleanup.clean(streamed.toString())
                    if (gen.get() == myGen) _translatedText.value = snapshot
                    true
                }
            } catch (t: Throwable) {
                Log.e(TAG, "Translation failed", t)
                // A transient generation failure shouldn't tear down the loaded model (that's
                // reserved for load failures). Surface it as an input error and let the user
                // retry; the model stays Ready.
                if (gen.get() == myGen) _inputError.value = t.message ?: "Translation failed."
            }
            if (gen.get() == myGen) _isTranslating.value = false
        }
    }

    fun cancelTranslation() {
        gen.incrementAndGet() // supersede the in-flight generation
        _isTranslating.value = false
    }

    fun clearAll() {
        cancelTranslation()
        _sourceText.value = ""
        _translatedText.value = ""
    }

    // MARK: - Editable inputs

    fun setSourceText(value: String) {
        _sourceText.value = value
    }

    fun setSource(language: Language) {
        _source.value = language
    }

    fun setTarget(language: Language) {
        _target.value = language
    }

    // MARK: - Language controls

    fun swapLanguages() {
        val oldSource = _source.value
        val oldTarget = _target.value

        if (oldSource.isDetect) {
            // Can't make "Detect" a target: promote the target to source, pick a sensible target.
            _source.value = oldTarget
            _target.value = if (oldTarget.id == "en") Language.named("ko") else Language.named("en")
        } else {
            _source.value = oldTarget
            _target.value = oldSource
        }

        // Mirror DeepL: move the produced translation up into the input, then re-run.
        if (_translatedText.value.isNotEmpty()) {
            _sourceText.value = _translatedText.value
            _translatedText.value = ""
            translate()
        }
    }

    // MARK: - Actions (offline-friendly)

    fun paste() {
        val clipboard = getApplication<Application>()
            .getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val text = clipboard.primaryClip?.takeIf { it.itemCount > 0 }
            ?.getItemAt(0)?.coerceToText(getApplication())?.toString()
        if (!text.isNullOrEmpty()) _sourceText.value = text
    }

    fun copyTranslation() {
        val clipboard = getApplication<Application>()
            .getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText("translation", _translatedText.value))
    }

    fun speak(text: String, language: Language) = speech.speak(text, language)

    // MARK: - Voice input (offline speech-to-text)

    /** Locale to recognize in: the source language, or the device locale when source is "Detect". */
    private fun recognitionLocale(): Locale =
        if (_source.value.isDetect) Locale.getDefault() else _source.value.speechLocale

    fun startVoiceInput() {
        if (_isListening.value) return
        _inputError.value = null
        _partialVoiceText.value = ""
        _isListening.value = true
        voice.start(
            locale = recognitionLocale(),
            onPartial = { _partialVoiceText.value = it },
            onFinal = { text -> finishVoice(text) },
            onError = { message ->
                _isListening.value = false
                _partialVoiceText.value = ""
                _inputError.value = message
            },
        )
    }

    /** Stop listening; a final transcript arrives via the controller's onFinal. */
    fun stopVoiceInput() = voice.stop()

    private fun finishVoice(text: String) {
        _isListening.value = false
        _partialVoiceText.value = ""
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return
        _sourceText.value = trimmed
        _showResult.tryEmit(Unit)
        translate()
    }

    // MARK: - Image OCR (offline)

    fun recognizeImage(uri: Uri) {
        _inputError.value = null
        _isRecognizingImage.value = true
        ImageTextRecognizer.recognize(
            context = getApplication(),
            uri = uri,
            language = _source.value,
            onResult = { text ->
                _isRecognizingImage.value = false
                val trimmed = text.trim()
                if (trimmed.isEmpty()) {
                    _inputError.value = "No text found in the image."
                    return@recognize
                }
                _sourceText.value = trimmed
                _showResult.tryEmit(Unit)
                translate()
            },
            onError = { message ->
                _isRecognizingImage.value = false
                _inputError.value = message
            },
        )
    }

    fun clearInputError() {
        _inputError.value = null
    }

    override fun onCleared() {
        super.onCleared()
        gen.incrementAndGet() // stop any in-flight generation
        engine.execute { translator.tearDown() } // release native model on the engine thread
        engine.shutdown()
        speech.shutdown()
        voice.close()
        networkMonitor.stop()
    }

    companion object {
        private const val TAG = "TranslationVM"
    }
}

File: apps/translate-tencent_HY-MT/iOS/OfflineTranslator/ViewModels/TranslationViewModel.swift
import Foundation
import SwiftUI
import AVFoundation
import os
#if canImport(UIKit)
import UIKit
#endif

@MainActor
final class TranslationViewModel: ObservableObject {
    // Editable translation state
    @Published var sourceText = ""
    @Published var translatedText = ""
    @Published var source: Language = .detect
    @Published var target: Language = .named("en")

    // Model + generation state
    @Published var modelState: ModelState = .loading
    @Published var downloadProgress: Double = 0
    @Published var isTranslating = false

    // Voice / OCR input state
    @Published var isListening = false
    @Published var partialVoiceText = ""
    @Published var isRecognizingImage = false
    @Published var inputError: String?
    /// Bumped when voice/OCR captures text, signalling the screen to show the result phase.
    @Published var showResultSignal = 0

    enum ModelState: Equatable {
        case loading
        case ready
        case failed(String)
    }

    private let translator: Translator
    private let synthesizer = AVSpeechSynthesizer()
    private let speechInput = SpeechInputController()

    // The ZeticMLange native model is blocking and not thread-safe: load + run +
    // waitForNextToken must all happen on ONE thread, never on Swift's cooperative
    // pool (Task.detached), which crashes the native init. A single dedicated serial
    // queue serializes every engine call — the pattern used by every working
    // ZeticMLange app. `genLock` holds the latest generation id so an in-flight
    // translation stops early when superseded.
    private let engineQueue = DispatchQueue(label: "ai.zetic.translate.engine", qos: .userInitiated)
    private let genLock = OSAllocatedUnfairLock(initialState: 0)
    private let log = Logger(subsystem: "ai.zetic.offlinetranslator", category: "model")

    var usingRealModel: Bool {
        #if canImport(ZeticMLange)
        return true
        #else
        return false
        #endif
    }

    init() {
        #if canImport(ZeticMLange)
        translator = ZeticTranslator()
        #else
        translator = MockTranslator()
        #endif
        loadModel()
    }

    // MARK: - Model lifecycle

    func loadModel() {
        modelState = .loading
        downloadProgress = 0
        let translator = self.translator
        let log = self.log
        log.info("Loading model \(ZeticConfig.modelName, privacy: .public) v\(ZeticConfig.modelVersion) …")
        engineQueue.async { [weak self] in
            do {
                try translator.load { progress in
                    DispatchQueue.main.async { self?.downloadProgress = progress }
                }
                log.info("Model ready")
                DispatchQueue.main.async { self?.modelState = .ready }
            } catch {
                let message = error.localizedDescription
                log.error("Model load failed: \(message, privacy: .public)")
                DispatchQueue.main.async { self?.modelState = .failed(message) }
            }
        }
    }

    // MARK: - Translation

    var canTranslate: Bool {
        modelState == .ready && !sourceText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    func translate() {
        let text = sourceText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, modelState == .ready else { return }

        // New generation id supersedes any in-flight one (its callback sees a newer id and stops).
        let myGen = genLock.withLock { v -> Int in v += 1; return v }
        translatedText = ""
        isTranslating = true

        let prompt = TranslationPrompt.make(text: text, from: source, to: target)
        let translator = self.translator
        let genLock = self.genLock
        engineQueue.async { [weak self] in
            translator.reset() // clear KV/conversation state from any prior turn
            var streamed = ""
            do {
                try translator.generate(prompt: prompt) { token in
                    guard genLock.withLock({ $0 }) == myGen else { return false } // superseded → stop
                    streamed += token
                    let snapshot = TranslationCleanup.clean(streamed)
                    DispatchQueue.main.async {
                        guard let self, genLock.withLock({ $0 }) == myGen else { return }
                        self.translatedText = snapshot
                    }
                    return true
                }
            } catch {
                // A transient generation failure shouldn't tear down the loaded model
                // (that's reserved for load failures). Surface it as an input error and
                // let the user retry; the model stays `.ready`.
                let message = error.localizedDescription
                DispatchQueue.main.async {
                    guard let self, genLock.withLock({ $0 }) == myGen else { return }
                    self.inputError = message
                }
            }
            DispatchQueue.main.async {
                guard let self, genLock.withLock({ $0 }) == myGen else { return }
                self.isTranslating = false
            }
        }
    }

    func cancelTranslation() {
        genLock.withLock { $0 += 1 } // supersede the in-flight generation
        isTranslating = false
    }

    func clearAll() {
        cancelTranslation()
        sourceText = ""
        translatedText = ""
    }

    // MARK: - Language controls

    func swapLanguages() {
        let oldSource = source
        let oldTarget = target

        if oldSource.isDetect {
            // Can't make "Detect" a target: promote the target to source and pick a sensible target.
            source = oldTarget
            target = oldTarget.id == "en" ? .named("ko") : .named("en")
        } else {
            source = oldTarget
            target = oldSource
        }

        // Mirror DeepL: move the produced translation up into the input, then re-run.
        if !translatedText.isEmpty {
            sourceText = translatedText
            translatedText = ""
            translate()
        }
    }

    // MARK: - Actions (offline-friendly)

    func paste() {
        #if canImport(UIKit)
        if let string = UIPasteboard.general.string {
            sourceText = string
        }
        #endif
    }

    func copyTranslation() {
        #if canImport(UIKit)
        UIPasteboard.general.string = translatedText
        #endif
    }

    /// Reads `text` aloud with on-device (offline) text-to-speech. Tapping again while
    /// speaking stops it. Configures the audio session so it's actually audible on a
    /// real device, and falls back to any installed voice for the language.
    func speak(_ text: String, language: Language) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }

        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
            return
        }

        // Without an active .playback session, TTS is silent on device (e.g. in
        // silent mode) or routes oddly. spokenAudio + duckOthers is the right mode.
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.playback, mode: .spokenAudio, options: [.duckOthers])
        try? session.setActive(true)

        let utterance = AVSpeechUtterance(string: trimmed)
        utterance.voice = bestVoice(for: language)
        synthesizer.speak(utterance)
    }

    /// Exact voice for the language code, else any installed voice sharing its prefix
    /// (e.g. "ko-…"), else the system default.
    private func bestVoice(for language: Language) -> AVSpeechSynthesisVoice? {
        let code = language.speechCode
        if let exact = AVSpeechSynthesisVoice(language: code) { return exact }
        let prefix = String(code.prefix(2))
        return AVSpeechSynthesisVoice.speechVoices().first { $0.language.hasPrefix(prefix) }
    }

    // MARK: - Voice input (offline speech-to-text)

    /// Locale to recognize in: the source language, or the device locale when source is "Detect".
    private var recognitionLocale: Locale {
        source.isDetect ? Locale.current : Locale(identifier: source.speechCode)
    }

    func startVoiceInput() {
        guard !isListening else { return }
        inputError = nil
        partialVoiceText = ""
        isListening = true
        speechInput.start(
            locale: recognitionLocale,
            onPartial: { [weak self] text in self?.partialVoiceText = text },
            onFinal: { [weak self] text in self?.finishVoice(text) },
            onError: { [weak self] message in
                self?.isListening = false
                self?.partialVoiceText = ""
                self?.inputError = message
            }
        )
    }

    /// Stop listening; the recognizer flushes a final transcript via the controller's onFinal.
    func stopVoiceInput() {
        speechInput.stop()
    }

    private func finishVoice(_ text: String) {
        isListening = false
        partialVoiceText = ""
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        sourceText = trimmed
        showResultSignal += 1
        translate()
    }

    // MARK: - Image OCR (offline)

    func recognize(image: UIImage) {
        inputError = nil
        isRecognizingImage = true
        let languages = source.isDetect ? [] : [source.speechCode]
        VisionTextRecognizer.recognize(image, languages: languages) { [weak self] result in
            guard let self else { return }
            self.isRecognizingImage = false
            switch result {
            case .success(let text):
                let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
                if trimmed.isEmpty {
                    self.inputError = "No text found in the image."
                    return
                }
                self.sourceText = trimmed
                self.showResultSignal += 1
                self.translate()
            case .failure(let error):
                self.inputError = error.localizedDescription
            }
        }
    }

    deinit {
        genLock.withLock { $0 += 1 } // stop any in-flight generation
        let translator = self.translator
        engineQueue.async { translator.tearDown() } // release native model on the engine thread
    }
}

/// Trims artifacts from streamed model output (leading whitespace, an accidental
/// echo of the prompt's lead-in, surrounding quotes).
enum TranslationCleanup {
    static func clean(_ raw: String) -> String {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        // If the model echoed the instruction line, drop it (Hunyuan rarely does, but be safe).
        for marker in ["Translate the following segment", "Translate the following text", "把下面的文本翻译成"] {
            if text.hasPrefix(marker), let newline = text.firstIndex(of: "\n") {
                text = String(text[text.index(after: newline)...]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        return text
    }
}
